### PR TITLE
KOGITO-3716 Allow automatic close of staging repo

### PR DESCRIPTION
**Pull request by radtriste**
https://issues.redhat.com/browse/KOGITO-3716

Dependent PRs:
- https://github.com/kiegroup/kogito-pipelines/pull/91
- https://github.com/kiegroup/kogito-runtimes/pull/865
- https://github.com/kiegroup/optaplanner/pull/1010
- https://github.com/kiegroup/kogito-examples/pull/435
- https://github.com/kiegroup/jenkins-pipeline-shared-libraries/pull/114

### CODE REVIEW

File: Jenkinsfile.deploy [hunk 1]
```diff
@@ -239,9 +250,35 @@ void mavenCleanInstall(String directory, boolean skipTests = false, List profile
...
+        }
+        runMaven('clean deploy', directory, true, [], extraArgs)
+    } else {
+        // First deploy locally
+        extraArgs = "-DaltDeploymentRepository=local::default::file://${env.MAVEN_DEPLOY_LOCAL_DIR}/${directory}"
+        runMaven('clean deploy', directory, true, [], extraArgs)
+
+        // Deploy to staging repository and close
+        cmd = 'org.sonatype.plugins:nexus-staging-maven-plugin:1.6.5:deploy-staged-repository -DkeepStagingRepositoryOnCloseRuleFailure=true -DstagingProgressTimeoutMinutes=10'
+        cmd += "--projects ${parentProjectName ?: directory}"
```
Comment (radtriste) [inline Jenkinsfile.deploy:267]:
```suggestion
        cmd += " --projects ${parentProjectName ?: directory}"
```

File: Jenkinsfile.deploy [hunk 2]
```diff
@@ -239,9 +250,35 @@ void mavenCleanInstall(String directory, boolean skipTests = false, List profile
...
+        }
+        runMaven('clean deploy', directory, true, [], extraArgs)
+    } else {
+        // First deploy locally
+        extraArgs = "-DaltDeploymentRepository=local::default::file://${env.MAVEN_DEPLOY_LOCAL_DIR}/${directory}"
+        runMaven('clean deploy', directory, true, [], extraArgs)
+
+        // Deploy to staging repository and close
+        cmd = 'org.sonatype.plugins:nexus-staging-maven-plugin:1.6.5:deploy-staged-repository -DkeepStagingRepositoryOnCloseRuleFailure=true -DstagingProgressTimeoutMinutes=10'
+        cmd += " --projects ${parentProjectName ?: directory}"
```
Comment (radtriste) [inline Jenkinsfile.deploy:267]:
```suggestion
        cmd += " --projects :${parentProjectName ?: directory}"
```

File: Jenkinsfile.deploy [hunk 3]
```diff
@@ -239,9 +250,35 @@ void mavenCleanInstall(String directory, boolean skipTests = false, List profile
...
+        cmd += " -DserverId=${env.NEXUS_RELEASE_REPOSITORY_ID}" 
+        cmd += " -DrepositoryDirectory=${env.MAVEN_DEPLOY_LOCAL_DIR}/${directory}"
+        cmd += " -DstagingProfileId=${env.NEXUS_STAGING_REPOSITORY_ID}" 
+        cmd += " -DstagingDescription=\"Kogito Runtimes ${getProjectVersion()}\""
+        echo "Run maven cmd $cmd"
+        runMaven(cmd, directory)
+
+        // Promote to build profile
+        cmd = "org.sonatype.plugins:nexus-staging-maven-plugin:1.6.5:promote -DbuildPromotionProfileId=${env.NEXUS_BUILD_PROMOTION_PROFILE_ID}"
+        cmd += "--projects ${directory}"
```
Comment (radtriste) [inline Jenkinsfile.deploy:278]:
```suggestion
        cmd += " --projects :${parentProjectName ?: directory}"
```

File: Jenkinsfile.deploy [hunk 4]
```diff
@@ -135,10 +150,6 @@ pipeline {
         stage('Build optaplanner') {
             steps {
                 script {
-                    String extraArgs = ''
-                    if (params.MAVEN_DEPENDENCIES_REPOSITORY != '') {
-                        extraArgs = "-s ${env.GLOBAL_MAVEN_SETTINGS} -Denforcer.skip"
-                    }
                     mavenCleanInstall('optaplanner', params.SKIP_TESTS, [], extraArgs)
```
Comment (radtriste) [inline Jenkinsfile.deploy:153]:
```suggestion
                    mavenCleanInstall('optaplanner', params.SKIP_TESTS)
```

**Comment by radtriste** [comment]
Almost there but testing is blocked by https://issues.redhat.com/browse/KOGITO-3795 ...

File: Jenkinsfile.promote [hunk 5]
```diff
@@ -181,44 +160,42 @@ pipeline {
...
-                        prLink = commitAndCreatePR("Release OptaPlanner ${getProjectVersion()}", prBranchName, 'master')
-                    }
-                    dir(websiteRepository) {
-                        checkoutRepo(websiteRepository, 'master')
-                        mergeAndPush(prLink, 'master')
-                    }
-                }
-            }
-        }
+        // TODO set back before merging
```
Comment (radtriste) [inline Jenkinsfile.promote:169]: Will be commented out once testing is finished

File: Jenkinsfile.promote [hunk 6]
```diff
@@ -478,50 +444,32 @@ void installGithubCLI() {
...
-        mvnCmd += " -DaltDeploymentRepository=runtimes-artifacts::default::${params.MAVEN_DEPLOY_REPOSITORY} -Denforcer.skip=true"
-    }
-
-    if (isSpecificMavenConfig()) {
-        runMaven(mvnCmd, true)
-    } else {
-        maven.runMavenWithSubmarineSettings(mvnCmd, true)
+void uploadDistribution(String directory) {
+    dir(directory) {
+        echo "uploadDistribution for ${directory}"
```
Comment (radtriste) [inline Jenkinsfile.promote:449]: Will be commented out once testing is finished

File: Jenkinsfile.deploy [hunk 7]
```diff
@@ -122,30 +141,20 @@ pipeline {
...
-                    mavenCleanInstallOptaPlannerParents('optaplanner')
-                    dir('optaweb-vehicle-routing') {
-                        maven.mvnVersionsUpdateParentAndChildModules(getProjectVersion())
-                    }
-                    dir('optaweb-employee-rostering') {
-                        maven.mvnVersionsUpdateParentAndChildModules(getProjectVersion())
-                    }
+                    maven.mvnVersionsSet(defaultOptaplannerMvnCmd, getProjectVersion())
+                    maven.mvnSetVersionProperty(defaultOptaplannerMvnCmd, 'version.org.kie.kogito', getKogitoVersion())
+                    defaultOptaplannerMvnCmd.clone().skipTests(true).withOptions(['-U']).run('clean install')
```
Comment (rsynek) [inline Jenkinsfile.deploy:146]: The `mavenCleanInstallOptaPlannerParents` ran only specific modules (a parent pom.xml and BOM) the other projects depend upon to save time (a couple of seconds vs. ~5 minutes).
I don't expect these two modules to be renamed; especially the BOM is a part of our API.

File: Jenkinsfile.deploy [hunk 8]
```diff
@@ -196,9 +209,23 @@ pipeline {
...
+                        runMavenDeploy(defaultOptawebEmployeeRosteringMvnCmd)
+                    } else {
+                        // Deploy locally and then to staging
+                        runMavenDeployLocally(defaultOptaplannerMvnCmd, optaplannerRepository)
+                        runMavenDeployLocally(defaultOptawebVehicleRoutingMvnCmd, vehicleRoutingRepository)
+                        runMavenDeployLocally(defaultOptawebEmployeeRosteringMvnCmd, employeeRosteringRepository)
+
+                        runMavenStage(defaultOptaplannerMvnCmd, optaplannerRepository)
+                        runMavenStage(defaultOptawebVehicleRoutingMvnCmd, vehicleRoutingRepository)
+                        runMavenStage(defaultOptawebEmployeeRosteringMvnCmd, employeeRosteringRepository)
```
Comment (rsynek) [inline Jenkinsfile.deploy:226]: The `if-else` can be replaced by a method, e.g. `deployMavenArtifacts()` that would take care of a single repository.

File: Jenkinsfile.deploy [hunk 9]
```diff
@@ -285,55 +310,37 @@ void commitAndCreatePRIgnoringNpmRegistry(String repo) {
...
-    while IFS= read -r pom; do
-        if ! git check-ignore -q "\$pom"; then
-            poms_to_add="\$poms_to_add \$pom"
-        fi
-    done < found_poms.txt
-    rm found_poms.txt
-    git add \$poms_to_add
-    git status
-    '''
-}
```
Comment (rsynek) [inline Jenkinsfile.deploy:303]: Thank you!

File: Jenkinsfile.deploy [hunk 10]
```diff
@@ -285,55 +310,37 @@ void commitAndCreatePRIgnoringNpmRegistry(String repo) {
...
+void runMavenDeploy(MavenCommand mvnCmd){
```
Comment (rsynek) [inline Jenkinsfile.deploy:313]:
```suggestion
void runMavenDeploy(MavenCommand mvnCmd) {
```

File: Jenkinsfile.deploy [hunk 11]
```diff
@@ -285,55 +310,37 @@ void commitAndCreatePRIgnoringNpmRegistry(String repo) {
...
+    mvnCmd= mvnCmd.clone()
+    
+    if(params.MAVEN_DEPLOY_REPOSITORY){
```
Comment (rsynek) [inline Jenkinsfile.deploy:316]:
```suggestion
    if(params.MAVEN_DEPLOY_REPOSITORY) {
```

File: Jenkinsfile.deploy [hunk 12]
```diff
@@ -285,55 +310,37 @@ void commitAndCreatePRIgnoringNpmRegistry(String repo) {
...
-    }
+MavenStagingHelper getStagingHelper(MavenCommand mvnCmd) {
+    return new MavenStagingHelper(this, mvnCmd)
+                .withNexusReleaseUrl(params.NEXUS_RELEASE_URL)
+                .withNexusReleaseRepositoryId(params.NEXUS_RELEASE_REPOSITORY_ID)
 }
 
-void updateKogitoVersion(String newVersion) {
-    maven.mvnSetVersionProperty('version.org.kie.kogito', newVersion)
+String getLocalDeploymentFolder(String localDeployId){
```
Comment (rsynek) [inline Jenkinsfile.deploy:342]:
```suggestion
String getLocalDeploymentFolder(String localDeployId) {
```

File: Jenkinsfile.deploy [hunk 13]
```diff
@@ -196,9 +209,23 @@ pipeline {
 
         stage('Deploy artifacts') {
             steps {
-                mavenDeploy('optaplanner', '-Dfull')
-                mavenDeploy('optaweb-vehicle-routing', "-s ${env.GLOBAL_MAVEN_SETTINGS}")
-                mavenDeploy('optaweb-employee-rostering', "-s ${env.GLOBAL_MAVEN_SETTINGS}")
+                script {
+                    if(!isRelease() || params.MAVEN_DEPLOY_REPOSITORY){
```
Comment (rsynek) [inline Jenkinsfile.deploy:213]:
```suggestion
                    if(!isRelease() || params.MAVEN_DEPLOY_REPOSITORY) {
```

File: Jenkinsfile.promote [hunk 14]
```diff
@@ -478,50 +443,32 @@ void installGithubCLI() {
...
+    return mvnCmd
 }
 
-void uploadDistribution() {
-    withCredentials(bindings: [sshUserPrivateKey(credentialsId: 'optaplanner-filemgmt',
-            keyFileVariable: 'SSH_KEY_JBOSS_FILEMGMT')]) {
-        // For testing, simulate connection via SSH:
-        // sh "ssh -i $SSH_KEY_JBOSS_FILEMGMT -oKexAlgorithms=+diffie-hellman-group1-sha1 optaplanner@filemgmt.jboss.org"
-        sh "./build/release/upload_distribution.sh ${getProjectVersion()} $SSH_KEY_JBOSS_FILEMGMT"
+void runMavenDeploy(MavenCommand mvnCmd){
```
Comment (rsynek) [inline Jenkinsfile.promote:468]:
```suggestion
void runMavenDeploy(MavenCommand mvnCmd) {
```

File: Jenkinsfile.deploy (hunk 8 above)
Comment (radtriste) [reply Jenkinsfile.deploy]: something like:
```
deployMavenArtifacts(defaultOptaplannerMvnCmd, optaplannerRepository)
deployMavenArtifacts(defaultOptawebVehicleRoutingMvnCmd, vehicleRoutingRepository)
deployMavenArtifacts(defaultOptawebEmployeeRosteringMvnCmd, employeeRosteringRepository)
```
and then
```
void deployMavenArtifacts(mvnCmd, directory){
    if(!isRelease() || params.MAVEN_DEPLOY_REPOSITORY){
         runMavenDeploy(mvnCmd)
    } else {
        runMavenDeployLocally(mvnCmd, directory)
        runMavenStage(mvnCmd, directory)
    }
}
```
?

File: Jenkinsfile.deploy (hunk 7 above)
Comment (radtriste) [reply Jenkinsfile.deploy]: woops. gonna correct that

Comment (radtriste) [reply Jenkinsfile.deploy]: updated

File: .gitignore [hunk 15]
```diff
@@ -11,6 +11,14 @@
 /*.ipr
 /*.iws
 /*.iml
+.classpath
+.project
+.settings
+.factorypath
```
Comment (rsynek) [inline .gitignore:17]: These are eclipse files? They should be already covered by `/.*` on the line 5.

File: .jenkins/tests/pom.xml [hunk 16]
```diff
@@ -0,0 +1,287 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (rsynek) [inline .jenkins/tests/pom.xml:2]: While I like the idea of having a PR check for jenkinsfiles, I have concerns about maintainability of this solution:
1/ it brings a lot of dependencies to the project (although it's a separate maven build)
2/ it duplicates the same (or very similar) setup in kogito-runtimes, kogito-examples and kogito-pipeline (in form of build.gradle) and maybe other repositories.

We should think about reducing the maintainability cost before merging the PRs.

First proposal: creating a parent pom.xml (in kogito-pipelines or jenkins-pipeline-shared-libraries ?) the other repositories would inherit from.

**Review by rsynek — CHANGES_REQUESTED** [review]
Please see my comment regarding the new pom.xml. Otherwise looks great.

File: .jenkins/tests/pom.xml (hunk 16 above)
Comment (radtriste) [reply .jenkins/tests/pom.xml:2]: I see your concerns. I will remove the tests and gonna implement it as part of another issue (I need those PRs to be merged asap)
and I note your proposal for later implementation :)

Comment (radtriste) [reply .jenkins/tests/pom.xml]: https://issues.redhat.com/browse/BXMSPROD-1125

Comment (rsynek) [reply .jenkins/tests/pom.xml]: Perfect, thank you!

**Review by rsynek — APPROVED** [review]

**Comment by radtriste** [comment]
@rsynek can you review one last time and merge if ok ?

**Review by rsynek — APPROVED** [review]

**Pull request merged**